### ISSUE TITLE
How can I configure two networks for ceph-docker cluster?

### ISSUE THREAD

**Issue by tangzhankun** [ceph/ceph-docker#147]
Hi,
I want to enable two networks using ceph-docker as production environment.
But current "entrypoint.sh" only create a very simple ceph.conf also no interface for adding cluster network configuration.
How can we support public and cluster network in ceph-docker?

I am not sure if there is a need for IPtable settings, but refer to this if you want:
http://ceph.com/docs/master/rados/configuration/network-config-ref/

**Comment by Ulexus** [comment]
You can specify the cluster network by passing the environment variable `${CEPH_CLUSTER_NETWORK}`.  Unfortunately, you are correct; it is not handled by anything for the `ceph.conf` generation.

To get this working with the KV-store, you would add it to the `globals` section... thought this is undocumented and likely to change.  See issue #146

**Comment by leseb** [comment]
Hum which image do you use?
If you ceph/daemon this done here: https://github.com/ceph/ceph-docker/blob/master/daemon/config.static.sh#L20-L21

So passing both `CEPH_CLUSTER_NETWORK` and `CEPH_PUBLIC_NETWORK` should be good.

As for the kv store, just populate the proper subnets just as referred here: https://github.com/ceph/ceph-docker/blob/master/examples/confd/populate.sh#L36-L37

Am I missing something?

**Comment by hookenz** [comment]
@leseb - I think the addition of a feature that can suck in a file that contains additional options and is merged with the one built during initialisation is needed.  It'll then cater for most peoples needs and future options.

Unfortunately it's not entirely simple as it needs to come from somewhere that the init process can read from like a bind-mount drive or read from stdin or something.

**Comment by tangzhankun** [comment]
hi @leseb,
thanks. I see. But I am using ceph/mon now.
adding these two lines in ceph.conf is enough? I saw the document of ceph. Seems ceph.conf should have "host" in it like this:
[osd.0]
        host = {hostname}

And IPtables described there, don't know if we should set IPtables. any idea?

BTW, Will ceph/mon, ceph/osd and ceph/radosgw be removed in the futue? If so, I'll change to ceph/daemon

**Comment by tangzhankun** [comment]
@hookenz 
Is it a must for ceph.conf to be a full version when bootstrapping the initial mon? I mean, could the daemons have different ceph.conf?
For instance, the first generated ceph.conf may have simplest configurations in it. But daemons launched later could insert additional configurations(maybe specific configurations for one OSD) to it? 
If so, we could only ensure that the initial mon can generated the common must-have configurations in ceph.conf. Then the user could add other configurations in it through different ways based on their use case.

**Comment by leseb** [comment]
@tangzhankun yes you only need those 2 lines in your `ceph.conf`. The `host` is not needed.
The `ceph/daemon` can easily take over your `ceph/mon`. So you can stop the `ceph/mon` and re-run it with the `ceph/daemon` image, the monitor will start again.

**Comment by tangzhankun** [comment]
hi, @leseb 
You mean only public and cluster network CIDR in ceph.conf? no daemon's host section needed?
oh, that's seems different from the ceph document.

**Comment by leseb** [comment]
@tangzhankun yes only public and cluster networks :).

**Comment by tangzhankun** [comment]
@leseb 
ok. that's cool. And is there a simple method to confirm that ceph cluster are running with two networks?

**Comment by leseb** [comment]
@tangzhankun well just verify that the process is listening on the good network range :).

**Comment by tangzhankun** [comment]
@leseb 
yes. Thanks very much! :+1: 
And have you considered that using ceph/daemon in Kubernetes? I suppose this is possible :)

**Comment by leseb** [comment]
@tangzhankun ceph/daemon works with k8s for the monitors already.
I need to do further testing for the remaining daemons.

**Comment by tangzhankun** [comment]
@leseb 
Hi,
I tried to use ceph/daemon to start a ceph cluster, but encountered permission error when launching OSD:
```sh
[root@ms0113 ~]# docker run --rm --pid=host --net=host --privileged --name osd0 -e JOURNAL_DIR=/var/lib/ceph/osd/journal -e OSD_TYPE=directory -v /root/ceph_config_root/etc/ceph:/etc/ceph:rw -v /root/ceph_config_root/var/lib/ceph/bootstrap-osd:/var/lib/ceph/bootstrap-osd:rw -v /root/ceph_config_root/var/lib/ceph/osd/ceph-0:/var/lib/ceph/osd/ceph-0:rw -v /root/ceph_config_root/var/lib/ceph/osd/jnl-0:/var/lib/ceph/osd/journal:rw ceph/daemon osd
Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
static: does not generate config
2015-08-31 02:25:07.269355 7fd4809ff900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-08-31 02:25:07.494651 7fd4809ff900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-08-31 02:25:07.511674 7fd4809ff900 -1 filestore(/var/lib/ceph/osd/ceph-0) could not find 23c2fcde/osd_superblock/0//-1 in index: (2) No such file or directory
2015-08-31 02:25:07.850941 7fd4809ff900 -1 created object store /var/lib/ceph/osd/ceph-0 journal /var/lib/ceph/osd/journal/journal.0 for osd.0 fsid 231c04e4-1a86-46eb-be16-94663eeedec2
HEALTH_WARN 64 pgs stuck inactive; 64 pgs stuck unclean
Error EACCES: access denied
```
I am not quite sure of my steps, I do it as follows:
* Create the directory tree in all hosts (say {somedir}/etc/ceph;{somedir}/var/lib/ceph/osd .. )
* Launch the initial monitor
```sh
docker run --rm --net=host --privileged --name mon0 -e MON_IP=172.16.1.78 -e CEPH_PUBLIC_NETWORK=172.16.0.0/16 -e CEPH_CLUSTER_NETWORK=10.1.0.0/16 -v /root/ceph_config_root/etc/ceph:/etc/ceph:rw -v /root/ceph_config_root/var/lib/ceph:/var/lib/ceph:rw ceph/daemon mon
```
* Scp all keyrings and config files to other hosts' corresponding directory
* run "docker exec mon0 ceph osd create" to register osd id
* In another host, launch the first OSD
```sh
mkdir -p /root/ceph_config_root/var/lib/ceph/osd/ceph-0
docker run --rm --pid=host --net=host --privileged --name osd0 -e OSD_TYPE=directory -v /root/ceph_config_root/etc/ceph:/etc/ceph:rw -v /root/ceph_config_root/var/lib/ceph/osd/ceph-0:/var/lib/ceph/osd/ceph-0:rw ceph/daemon osd
```
I can launch successfully when using ceph/mon and ceph/osd. Here it reported the permission issue. I know ceph/daemon may need more authentication, but I copied the ceph.keyring. Any idea?

**Comment by leseb** [comment]
Hey @tangzhankun I'm not sure between those 2 lines which one is failing:

* https://github.com/ceph/ceph-docker/blob/master/daemon/entrypoint.sh#L201
* https://github.com/ceph/ceph-docker/blob/master/daemon/entrypoint.sh#L204

I assume it's the first one, can you run a bash a container and then try to run the command from the first one and the other one as well?

**Comment by tangzhankun** [comment]
@leseb 
Hi, It's caused by L204.
I tried the two command using a ceph/base image, it failed on L204:
```sh
[root@ms0114 ~]# docker run -it --rm --pid=host --net=host --name osd0 -e OSD_TYPE=directory -v /root/ceph_config_root/etc/ceph:/etc/ceph:rw -v /root/ceph_config_root/var/lib/ceph/osd/ceph-0:/var/lib/ceph/osd/ceph-0:rw -v /root/ceph_config_root/var/lib/ceph/bootstrap-osd:/var/lib/ceph/bootstrap-osd:rw ceph/base
root@ms0114:/# ceph -v
ceph version 0.94.2 (5fb85614ca8f354284c713a2f9c610860720bbf3)
root@ms0114:/# timeout 10 ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring health || exit 1
HEALTH_WARN 64 pgs stuck inactive; 64 pgs stuck unclean
root@ms0114:/# ceph -v
ceph version 0.94.2 (5fb85614ca8f354284c713a2f9c610860720bbf3)
root@ms0114:/# ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring auth get-or-create osd.0 osd 'allow *' mon 'allow profile osd' -o /var/lib/ceph/osd/ceph-0/keyring
Error EACCES: access denied
```
And see logs from the monitor as below:
```sh
2015-09-01 00:51:23.345147 7f4863c4f700  0 mon.ms0112@0(leader) e1 handle_command mon_command({"prefix": "health"} v 0) v1
2015-09-01 00:51:23.345179 7f4863c4f700  0 log_channel(audit) log [DBG] : from='client.? 172.16.1.80:0/1230591' entity='client.bootstrap-osd' cmd=[{"prefix": "health"}]: dispatch
2015-09-01 00:51:23.592440 7f4863c4f700  0 mon.ms0112@0(leader) e1 handle_command mon_command({"prefix": "auth get-or-create", "entity": "osd.0", "caps": ["osd", "allow *", "mon", "allow profile osd"]} v 0) v1
2015-09-01 00:51:23.592479 7f4863c4f700  1 mon.ms0112@0(leader) e1 handle_command access denied
2015-09-01 00:51:23.592494 7f4863c4f700  0 log_channel(audit) log [INF] : from='client.? 172.16.1.80:0/1230622' entity='client.bootstrap-osd' cmd=[{"prefix": "auth get-or-create", "entity": "osd.0", "caps": ["osd", "allow *", "mon", "allow profile osd"]}]:  access denied
```
This can also direct us to the L204. any idea?

**Comment by Ulexus** [comment]
Is it possible that the bootstrap keys you have are not the same ones the monitor has... or, perhaps, that you don't have them loaded on the monitor?

Try running a `ceph auth list` command on a working monitor and see if the bootstrap-osd keys match (and that the permissions on the monitor for that key are correct:  "[mon] allow profile bootstrap-osd")

**Comment by tangzhankun** [comment]
@Ulexus 
No. I am using the same keys. I tried launch OSD in the same host with Monitor, still same error
```sh
[root@ms0112 ~]# docker exec mon0 ceph auth list
installed auth entries:

client.admin
	key: AQBXZuZVraTvDhAALimVfEvz9bIuOlc8Up/gZA==
	auid: 0
	caps: [mds] allow
	caps: [mon] allow *
	caps: [osd] allow *
client.bootstrap-mds
	key: AQBXZuZVdngeEBAATa1CwLQ37C1CfL6vhUTVRA==
	caps: [mon] allow profile bootstrap-mds
client.bootstrap-osd
	key: AQBXZuZVmyOZDxAAmJoFQI9crDgB8vYy3ZpEHA==
	caps: [mon] allow profile bootstrap-osd
client.bootstrap-rgw
	key: AQBXZuZV0SuOEBAAiHa6b4AiFNnXeLZK5iOJbw==
	caps: [mon] allow profile bootstrap-rgw
[root@ms0112 ~]# docker exec mon0 ceph osd create
0
[root@ms0112 ~]# docker run --rm --pid=host --net=host --privileged --name osd0 -e OSD_TYPE=directory -v /root/ceph_config_root/etc/ceph:/etc/ceph:rw -v /root/ceph_config_root/var/lib/ceph/osd/ceph-0:/var/lib/ceph/osd/ceph-0:rw -v /root/ceph_config_root/var/lib/ceph/bootstrap-osd:/var/lib/ceph/bootstrap-osd:rw ceph/daemon osd
Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
static: does not generate config
2015-09-02 03:01:54.767202 7f8704ace900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-09-02 03:01:55.026242 7f8704ace900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-09-02 03:01:55.027035 7f8704ace900 -1 filestore(/var/lib/ceph/osd/ceph-0) could not find 23c2fcde/osd_superblock/0//-1 in index: (2) No such file or directory
2015-09-02 03:01:55.326199 7f8704ace900 -1 created object store /var/lib/ceph/osd/ceph-0 journal /var/lib/ceph/osd/ceph-0/journal for osd.0 fsid fd249d36-f1a8-47ed-ae18-7e7a511dca67
HEALTH_WARN 64 pgs stuck inactive; 64 pgs stuck unclean
Error EACCES: access denie
```

**Comment by Ulexus** [comment]
The above output does not prove you are using the same keys.  You may have already checked the keys, but since you didn't say so, I will assume that the evidence you have to support the claim that they are the same is that you executed the OSD on the same machine as the mon and received the same problem.  Please verify the key stored within `/root/ceph_config_root/var/lib/ceph/bootstrap-osd` matches the key you acquired from `ceph auth list` (and copied above).

Sorry for being pedantic; I just need clear data.

**Comment by tangzhankun** [comment]
@Ulexus 
Sorry for not that clear. Below is the "cat /root/ceph_config_root/var/lib/ceph/bootstrap-osd":
```sh
[root@ms0112 ~]# cat /root/ceph_config_root/var/lib/ceph/bootstrap-osd/ceph.keyring 
[client.bootstrap-osd]
	key = AQBXZuZVmyOZDxAAmJoFQI9crDgB8vYy3ZpEHA==
	caps mon = "allow profile bootstrap-osd"
```
and ceph auth list:
```sh
[root@ms0112 ~]# docker exec mon0 ceph auth list
installed auth entries:

client.admin
	key: AQBXZuZVraTvDhAALimVfEvz9bIuOlc8Up/gZA==
	auid: 0
	caps: [mds] allow
	caps: [mon] allow *
	caps: [osd] allow *
client.bootstrap-mds
	key: AQBXZuZVdngeEBAATa1CwLQ37C1CfL6vhUTVRA==
	caps: [mon] allow profile bootstrap-mds
client.bootstrap-osd
	key: AQBXZuZVmyOZDxAAmJoFQI9crDgB8vYy3ZpEHA==
	caps: [mon] allow profile bootstrap-osd
client.bootstrap-rgw
	key: AQBXZuZV0SuOEBAAiHa6b4AiFNnXeLZK5iOJbw==
	caps: [mon] allow profile bootstrap-rgw
```
and I bind-mounted the directoried that initial mon generated to run the OSD. I guess they cannot be different.
For further double check, I launched a ceph/base image to check the keys from inside the container
```sh
[root@ms0112 ~]# docker run -it --rm --pid=host --net=host --privileged --name osd0 -e OSD_TYPE=directory -v /root/ceph_config_root/etc/ceph:/etc/ceph:rw -v /root/ceph_config_root/var/lib/ceph/osd/ceph-0:/var/lib/ceph/osd/ceph-0:rw -v /root/ceph_config_root/var/lib/ceph/bootstrap-osd:/var/lib/ceph/bootstrap-osd:rw ceph/base
Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
                                                       root@ms0112:/# 
root@ms0112:/# 
root@ms0112:/# ceph -v
ceph version 0.94.2 (5fb85614ca8f354284c713a2f9c610860720bbf3)
root@ms0112:/# cat /var/lib/ceph/bootstrap-osd/ceph.keyring 
[client.bootstrap-osd]
	key = AQBXZuZVmyOZDxAAmJoFQI9crDgB8vYy3ZpEHA==
	caps mon = "allow profile bootstrap-osd"
```
So I think the keys are the same.

**Comment by Ulexus** [comment]
Indeed, that confirms they are the same.  Thanks.

I don't have any ideas at the moment, but I'll see if I can set up a similar test and replicate the problem.

**Comment by leseb** [comment]
I can replicate the problem. Give me some time to investigate.

**Comment by leseb** [comment]
We probably should use the `client.admin` as the other key hasn't enough permission.

**Comment by Ulexus** [comment]
That's fine with me, but isn't that the point of the bootstrap keys? (Please pardon my ignorance on the bootstrap keys)

**Comment by leseb** [comment]
@Ulexus yes yes this is just me, coming with a fix in a second :)

**Comment by tangzhankun** [comment]
@leseb 
I tried pull ceph/daemon from docker.io. But got this error:
```sh
[root@ms0112 ~]# docker run --rm --pid=host --net=host --privileged --name osd0 -e OSD_TYPE=directory -v /root/ceph_config_root/etc/ceph:/etc/ceph:rw -v /root/ceph_config_root/var/lib/ceph/osd/ceph-0:/var/lib/ceph/osd/ceph-0:rw -v /root/ceph_config_root/var/lib/ceph/bootstrap-osd:/var/lib/ceph/bootstrap-osd:rw ceph/daemon osd
Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
static: does not generate config
2015-09-07 05:20:19.528867 7f21f94f3900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-09-07 05:20:19.872914 7f21f94f3900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-09-07 05:20:19.880778 7f21f94f3900 -1 filestore(/var/lib/ceph/osd/ceph-0) could not find 23c2fcde/osd_superblock/0//-1 in index: (2) No such file or directory
2015-09-07 05:20:20.221916 7f21f94f3900 -1 created object store /var/lib/ceph/osd/ceph-0 journal /var/lib/ceph/osd/ceph-0/journal for osd.0 fsid 48965dbd-a074-41cb-9d85-e3fd3ef225fb
2015-09-07 05:20:20.221980 7f21f94f3900 -1 auth: error reading file: /var/lib/ceph/osd/ceph-0/keyring: can't open /var/lib/ceph/osd/ceph-0/keyring: (2) No such file or directory
2015-09-07 05:20:20.222177 7f21f94f3900 -1 created new key in keyring /var/lib/ceph/osd/ceph-0/keyring
HEALTH_WARN 64 pgs stuck inactive; 64 pgs stuck unclean
added key for osd.0
Invalid command:  root=default doesn't represent a float
osd crush create-or-move <osdname (id|osd.id)> <float[0.0-]> <args> [<args>...] :  create entry or move existing entry for <name> <weight> at/to location <args>
Error EINVAL: invalid command
```
Seems error from this line:
```sh
ceph ${CEPH_OPTS} --name=osd.${OSD_ID} --keyring=/var/lib/ceph/osd/${CLUSTER}-${OSD_ID}/keyring osd crush create-or-move -- ${OSD_ID} ${OSD_WEIGHT} ${CRUSH_LOCATION}
```
Is the patch passed testing? or my image from docker.io is not latest?

**Comment by tangzhankun** [comment]
@leseb 
I fix this by add -e OSD_WEIGHT=1.0 to my docker run command. Your patch works.
But shouldn't this OSD_WEIGHT has a default value 1.0? I don't have to set this in ceph/osd.
And the document said it have this default value. This is an inconsistency.

**Comment by leseb** [comment]
@tangzhankun #151 just fixed this issue, sorry about that.
